ocspRequest v1

KamilaErmolaeva opened the pull request.

Commit:
ocspRequest v1

microshine commented on src/ocsp/cert_id.ts:
```diff
@@ -0,0 +1,67 @@
+  protected onInit(asn: ocsp.CertID): void {
+    const algProv = container.resolve<AlgorithmProvider>(diAlgorithmProvider);
+    this.algorithm = algProv.toWebAlgorithm(asn.hashAlgorithm) as HashedAlgorithm;
+    this.issuerNameHash = AsnConvert.serialize(asn.issuerNameHash);
+    this.issuerKeyHash = AsnConvert.serialize(asn.issuerKeyHash);
+    this.serialNumber = AsnConvert.toString(AsnConvert.serialize(asn.serialNumber));
```
Меня смущают преобразования через `AsnConver.serialize`. Если я не ошибаюсь, то `asn.issuerNameHash` и `asn.issuerKeyHash` уже имеют байтовое представление и его можно извлечь через свойство `buffer`

```diff
- this.issuerNameHash = AsnConvert.serialize(asn.issuerNameHash);
- this.issuerKeyHash = AsnConvert.serialize(asn.issuerKeyHash);
+ this.issuerNameHash = asn.issuerNameHash.buffer;
+ this.issuerKeyHash = asn.issuerKeyHash.buffer;
```

Пример преобразования `serialNumber`  нужно взять из `X509Certificate`

```ts
this.serialNumber = Convert.ToHex(asn.serialNumber);
```

microshine commented on src/ocsp/ocsp_request.ts:
```diff
@@ -0,0 +1,241 @@
+
+  protected onInit(asn: ocsp.Request): void {
+    const algProv = container.resolve<AlgorithmProvider>(diAlgorithmProvider);
+    this.certificateID.algorithm = algProv.toWebAlgorithm(asn.reqCert.hashAlgorithm) as HashedAlgorithm;
+    this.certificateID.issuerKeyHash = AsnConvert.serialize(asn.reqCert.issuerKeyHash);
+    this.certificateID.issuerNameHash = AsnConvert.serialize(asn.reqCert.issuerNameHash);
+    this.certificateID.serialNumber = AsnConvert.toString(AsnConvert.serialize(asn.reqCert.serialNumber));
```
Тут идет повторная реализация чтения `CertificateID`. Достаточно было использовать уже существующий класс

```ts
this.certificateID = new CertificateID(asn.reqCert);
```

Commit:
tests ocsp request

microshine commented on src/ocsp/cert_id.ts:
```diff
@@ -0,0 +1,82 @@
+   * @param serialNumber Hexadecimal string of the serial number
+   * @param crypto Crypto provider. Default is from CryptoProvider
+   */
+  public static async create(algorithm: AlgorithmIdentifier, issuer: X509Certificate, serialNumber: string): Promise<CertificateID> {
```
Неправильная реализация. `issuerKeyHash` и `issuerNameHash` должны хранить результаты `digest` функции. На основе заданного `algorithm` нужно получить хэш из соответствующих полей сертификата `issuer`.  В описании параметров указан параметр `crypto`, но он отсутствует в реализации

microshine commented on src/ocsp/cert_id.ts:
```diff
@@ -0,0 +1,82 @@
+import { container } from "tsyringe";
+import { AlgorithmProvider, diAlgorithmProvider } from "../algorithm";
+import { Convert } from "pvtsutils";
+import { OctetString } from "@peculiar/asn1-schema";
```
Импорты внешних модулей (`tsyringe`, `pvtsutils`, `@peculiar/asn1-schema`) должны происходить раньше импортов локальных файлов

microshine commented on src/ocsp/ocsp_request.ts:
```diff
@@ -0,0 +1,237 @@
+   * @remarks
+   * If the OCSP request is not signed, then the check always returns `true`.
+   */
+  public async verify(signer: PublicKeyType, crypto: Crypto): Promise<boolean> {
```
Сделать `crypto` опциональным. Использовать default провайдер по умолчанию (как в других подобных функциях)

microshine commented on src/ocsp/ocsp_request.ts:
```diff
@@ -0,0 +1,237 @@
+import { IAsnSignatureFormatter, diAsnSignatureFormatter } from "../asn_signature_formatter";
+import { PemConverter } from "../pem_converter";
+import { AlgorithmProvider, diAlgorithmProvider } from "../algorithm";
+import { AsnConvert } from "@peculiar/asn1-schema";
```
Исправить импорт модулей. Сначала внешние модули.

microshine commented on src/ocsp/ocsp_request_generator.ts:
```diff
@@ -0,0 +1,106 @@
+import { AlgorithmProvider, diAlgorithmProvider } from "../algorithm";
+import { cryptoProvider } from "../provider";
+import { IAsnSignatureFormatter, diAsnSignatureFormatter } from "../asn_signature_formatter";
+import { AsnConvert, OctetString } from "@peculiar/asn1-schema";
```
Исправить импорт модулей.

microshine commented on src/ocsp/ocsp_request_generator.ts:
```diff
@@ -0,0 +1,106 @@
+      }
+    }
+
+    return new OCSPRequest(AsnConvert.serialize(asnOcspReq));
```
Для чего тут `serialize`? Вроде как можно создать OCSPRequest из asnOcspReq

```js
new OCSPRequest(asnOcspReq)
```

microshine commented on test/test_x509_ocsp.ts:
```diff
@@ -0,0 +1,46 @@
+Cuv/UlaFeXwPEDtbvhgY8FIBW8CNRLyA
+-----END CERTIFICATE-----`;
+
+  it("generating an OCSP request", async () => {
```
Нужно добавить тесты:

- создание OCSP запроса с подписью
- проверка подписи OCSP запроса
- чтение OCSP запроса

microshine replied:
Вот тут можно посмотреть как сгенерировать запрос на сертификат с помощью OpenSSL. Что даст возможность сопоставить результаты нашего модуля и OpenSSL

https://www.e2encrypted.com/pki/pki-certificate-revoke-ocsp/

microshine commented on src/ocsp/ocsp_request.ts:
```diff
@@ -0,0 +1,237 @@
+import * as ocsp from "@peculiar/asn1-ocsp";
```
Нжно добавить пакет в зависимости проекта

```
Cannot find module '@peculiar/asn1-ocsp' or its corresponding type declarations.ts(2307)
```

microshine replied:
```sh
yarn add @peculiar/asn1-ocsp
```

microshine commented on src/ocsp/index.ts:
Необходимо поднять процент тестов

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/7333669/236134327-72f183ba-85bd-4690-9807-ccac54e6ef67.png">

Используй команду `npm run coverage` для генерации отчета. Отчет в формате HTML `coverage/lcov-report/index.html`

Commit:
fix ocsp request

microshine commented on src/ocsp/ocsp_request_generator.ts:
```diff
@@ -0,0 +1,106 @@
+   * @returns OCSP request
+   */
+  public static async create(params: OCSPRequestCreateParams, crypto = cryptoProvider.get()): Promise<OCSPRequest> {
+    const certID = await CertificateID.create(params.certificate.signatureAlgorithm.hash, params.issuer, params.certificate.serialNumber);
```
Нужно добавить параметр в `OCSPRequestCreateParams` - `hashAlgorithm`. И использовать его для `CertificateID.create(params.hashAlgorithm`. Параметр должен быть опциональным. Значение должно быть опциональным, если его нет, то `SHA-256`

```js
/**
 * The hashing algorithm identifier for CertificateID generation. Default is SHA-256.
 */
hashAlgorithm?: AlgorithmIdentifier;
```

Commit:
ocsp_response

yaz9900 committed:
Fix for OCSP Request Generation

Fixed the generation of CertID to generate correct hash of public key and OID for the hash function.
Fixed location of the nonce extension.
Extended text_x509_ocsp.ts to test ocsp request on microsoft certificate and validate with ocsp server.
added @peculiar/asn1-ocsp dependency to yarn.

yaz9900 committed:
Authority Information Access Extension

Created Authority Information Access Extension to properly parse OCSP server and certificate address of OCSP server. Extended test_x509_ocsp.ts to pull OCSP link out of certificate

microshine commented on src/extensions/authority_information_access.ts:
```diff
@@ -0,0 +1,45 @@
+  }
+
+
+
```
Удалить лишние строки

microshine commented on src/extensions/authority_information_access.ts:
```diff
@@ -0,0 +1,45 @@
+ */
+export class AuthorityInformationAccessExtension extends Extension {
+
+
```
Удалить лишние строки

microshine commented on src/extensions/authority_information_access.ts:
```diff
@@ -0,0 +1,45 @@
+import { BufferSourceConverter } from "pvtsutils";
+import { Extension } from "../extension";
+import { GeneralNames, JsonGeneralNames } from "../general_name";
+import { TextObject } from "../text_converter";
```
Удалить неиспользуемые переменные

microshine commented on src/extensions/authority_information_access.ts:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Represents the Authority Information Access certificate extension
+ */
+export class AuthorityInformationAccessExtension extends Extension {
```
Необходимо изменить тип для `AuthorityInfoAccessType`, чтобы он не использовал типы из `@peculiar/asn1-x509`. В качестве примера вы можете обратиться к реализации `SubjectAlternativeNameExtension`, которая использует `GeneralNames` из этого модуля. Также я заметил, что нет тестов для работы с `AuthorityInformationAccessExtension`. Пожалуйста, убедитесь, что вы добавили соответствующие модульные тесты для проверки функциональности `AuthorityInformationAccessExtension` после внесения изменений.

microshine commented on test/test_x509_ocsp.ts:
```diff
@@ -0,0 +1,130 @@
+import * as assert from "assert";
+import * as x509 from "../src";
+import { Convert } from "pvtsutils";
```
Удалить неиспользуемые переменные

microshine commented on test/test_x509_ocsp.ts:
```diff
@@ -0,0 +1,130 @@
+    for(const accessDescription of authInfoAccess.data){
+      if(typeof(accessDescription) === "string") throw new Error("AccessDescription is not an object");
+      if(accessDescription.accessMethod === "1.3.6.1.5.5.7.48.1"){
+        var ocspURL = accessDescription.accessLocation.uniformResourceIdentifier;
```
Исправить ошибку eslint. Необходимо использовать `let` или `const`

microshine commented on test/test_x509_ocsp.ts:
```diff
@@ -0,0 +1,130 @@
+    const response = await fetch(ocspURL, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/ocsp-request',
```
Исправить eslint ошибку. Использовать двойные кавычки `"`

microshine commented on src/extensions/authority_information_access.ts:
```diff
@@ -0,0 +1,45 @@
+
+
+
+}
```
Необходимо добавить два новых метода в класс `AuthorityInformationAccessExtension`. Эти методы должны упрощать получение значений `caIssuers` и `ocsp` из данных расширения.

```ts
export class AuthorityInformationAccessExtension extends Extension {
  // ...

  /**
   * Returns the CA Issuers from the extension data
   */
  public getCaIssuers(): string[] {
    // Implementation here
  }

  /**
   * Returns the OCSP from the extension data
   */
  public getOcsp(): string[] {
    // Implementation here
  }
}
```

microshine commented on test/test_x509_ocsp.ts:
```diff
@@ -0,0 +1,130 @@
+      body: postRequest,
+    });
+
+    if (!response.ok) {
```
использовать `assert.ok` или `assert.equal`

microshine commented on test/test_x509_ocsp.ts:
```diff
@@ -0,0 +1,130 @@
+
+    const data = await response.arrayBuffer();
+    // console.log(Convert.ToString(data, 'hex')); // Logs the ArrayBuffer as a Uint8Array for readability
+    assert.notEqual(data.byteLength, 5, "OCSP response is malformed")
```
Странная проверка результата. Думаю тут стоит читать полученный ответ черех `OCSPRepsonse` и проверять статус внутри сообщения.

yaz9900 committed:
Cleanup and Authority info access extension test

Cleaned up the code and added test for Authority info access extension

yaz9900 committed:
Added getCaIssuers() and getOcsp() to authority information access extension

yaz9900 committed:
first implementation of OCSP verification rule

first implementation of OCSP verification rule, reworked rules to use enum for type, added rule verification data to node tree. reversed the direction of the rule pass to root certificate to the leaf.

microshine commented on src/ocsp/cert_id.ts:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    const hashAlgorithm  = new asn1X509.AlgorithmIdentifier({ algorithm: algorithm, parameters: null })
+    hashAlgorithm.algorithm = "1.3.14.3.2.26";
```
Лучше использовать провайдер алгоритмов для формирования `AlgorithmIdentifier` ([пример](https://github.com/PeculiarVentures/x509/blob/ecf78224fd594abbc2fa83c41565d79874f88e00/src/pkcs10_cert_req_generator.ts#L93-L95))

```js
const algProv = container.resolve<AlgorithmProvider>(diAlgorithmProvider);
const hashAlgorithm = algProv.toAsnAlgorithm(signingAlgorithm);
```

microshine commented on src/ocsp/ocsp_response_generator.ts:
```diff
@@ -0,0 +1,152 @@
+      signatureAlgorithm: algProv.toAsnAlgorithm(params.signatureAlgorithm as HashedAlgorithm)
+    });
+
+    if (params.certificates) {
+      for (const cert of params.certificates) {
+        let spki: BufferSource;
+        if (cert.publicKey instanceof PublicKey) {
+          spki = cert.publicKey.rawData;
+        } else if (BufferSourceConverter.isBufferSource(cert.publicKey)) {
+          spki = cert.publicKey;
+        } else {
+          spki = await crypto.subtle.exportKey("spki", cert.publicKey);
+        }
+
+        const asnX509 = new asn1X509.Certificate({
+          tbsCertificate: new asn1X509.TBSCertificate({
+            version: asn1X509.Version.v3,
+            serialNumber: Convert.FromHex(cert.serialNumber),
+            signature: algProv.toAsnAlgorithm(params.signatureAlgorithm as HashedAlgorithm),
+            validity: new asn1X509.Validity({
+              notBefore: cert.notBefore,
+              notAfter: cert.notAfter,
+            }),
+            subject: AsnConvert.parse(cert.subjectName.toArrayBuffer(), asn1X509.Name),
+            issuer: AsnConvert.parse(cert.issuerName.toArrayBuffer(), asn1X509.Name),
+            extensions: new asn1X509.Extensions(cert.extensions.map(o => AsnConvert.parse(o.rawData, asn1X509.Extension)) || []),
+            subjectPublicKeyInfo: AsnConvert.parse(spki, asn1X509.SubjectPublicKeyInfo),
+          }),
+          signatureAlgorithm: algProv.toAsnAlgorithm(params.signatureAlgorithm as HashedAlgorithm),
+          signatureValue: signatureValue,
+        });
+
+        basicOCSPResp.certs?.push(asnX509);
+      }
```
Думаю что тут мы должны просто добавить сертификаты переданные в параметре, а не должны создавать сертификаты из параметров. Что-то вроде

```js
if (params.certificates) {
  basicOCSPResp.certs = [...params.certificates]
}
```

microshine commented on src/ocsp/ocsp_request_generator.ts:
```diff
@@ -0,0 +1,113 @@
+    });
+
+    if (params.requestorName) {
+      asnOcspReq.tbsRequest.requestorName = params.requestorName as asn1X509.GeneralName;
```
Не думаю что это правильное преобразование. Взможно при использовании тут будет ошибка. Думаю, что тут нужно использовать `params.requestorName.rawData`чтобы сформировать `asn1X509.GeneralName`

```js
asnOcspReq.tbsRequest.requestorName = AsnConvert.parse(params.requestorName.rawData, asn1X509.GeneralName);
```

Было бы хорошо добавить тест, чтобы отработать такой случай

microshine commented on src/ocsp/ocsp_request_generator.ts:
```diff
@@ -0,0 +1,113 @@
+        throw Error("Cannot convert ASN.1 signature value to WebCrypto format");
+      }
+
+      if (asnOcspReq.optionalSignature) {
+        asnOcspReq.optionalSignature.signature = asnSignature;
+        asnOcspReq.optionalSignature.signatureAlgorithm = algProv.toAsnAlgorithm(params.signatureAlgorithm as HashedAlgorithm);
+      }
```
тут нужно создать и заполнить `asnOcspReq.optionalSignature`, иначе мы всенда будем попадать на исключение, т.к. `asnOcspReq.optionalSignature` не инициализируется.

Добавить тест с созданием `optionalSignature`

microshine commented on src/ocsp/basic_ocsp_response.ts:
```diff
@@ -0,0 +1,178 @@
+  public findResponder(): X509Certificate | null {
+    if (this.responderID) {
+
+      const cert = typeof this.responderID === "string" ?
+        new X509Certificate(PemData.toArrayBuffer(this.responderID))
+        : new X509Certificate(this.responderID);
```
Тут ошибка. Сертификат не может быть создан из `responderID`. Необхо искать сертификат серди приложенных сертификатов сертификатов или использовать подключенное хранилище `certificateStorage`

microshine commented on src/ocsp/basic_ocsp_response.ts:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  /**
+   * Search for the certificate for which the status is being requested
```
нестыковка в описании, так как ищется не сертификат, а `SingleResponse`

microshine commented on src/ocsp/basic_ocsp_response.ts:
```diff
@@ -0,0 +1,178 @@
+   * @remarks
+   * Checks only the mathematical correctness of the signature, does not check the respondent's certificate.
+   */
+  public async verify(crypto = cryptoProvider.get()): Promise<boolean> {
+    const publicKey = await new PublicKey(this.tbs).export(crypto);
+
+    return await crypto.subtle.verify(this.signatureAlgorithm, publicKey, this.signature, this.tbs);
+  }
```
Для проверки нужны дополнительные параметры по аналогии с реализацией `verify` в OCSPRequest. Также использовать схожую реализацию.

https://github.com/KamilaErmolaeva/x509/blob/b3149e01fa9b516bc77e1302b78a78804898d895/src/ocsp/ocsp_request.ts#L185

Если параметры не указаны, то нужно получить улюч проверки через `findResponder`

Добавить тесты:
- проверка с использованием параметров метода
- проверка когда `Responder` присутствует в списке сертификатов `BasicOCSPResponse`
- проверка когда `Responder` присутствует в `certificateStorage`

microshine reviewed (changes requested):
Требуются правки

yaz9900 committed:
Cert ID fix

yaz9900 committed:
Delete package.json

yaz9900 committed:
Add asn1-schema, add null to algo
